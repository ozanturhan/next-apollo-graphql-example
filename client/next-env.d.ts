/// <reference types="next" />
/// <reference types="next/types/global" />

declare module '*.svg' {
  const value: string;
  export = value;
}
