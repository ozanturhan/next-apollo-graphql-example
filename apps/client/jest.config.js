module.exports = {
  coverageDirectory: '../../coverage/apps/client',
  testRegex: '(/__test__/.*|\\.(test|spec))\\.(ts|tsx|js|jsx)$',
  moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json'],
  moduleNameMapper: {
    '@components': '<rootDir>/components',
    '@queries': '<rootDir>/graphql/queries',
    '\\.(css)$': '<rootDir>/__mocks__/styleMock.js',
    '\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$':
      '<rootDir>/__mocks__/fileMock.js',
  },
  transform: {
    '^.+\\.(js|jsx|ts|tsx)$': 'babel-jest',
  },
  setupFiles: ['./setupJest.ts'],
};
