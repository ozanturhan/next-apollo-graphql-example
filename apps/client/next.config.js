const withImages = require('next-images');
module.exports = withImages();

module.exports = {
  distDir: '../../dist/.next',
};
