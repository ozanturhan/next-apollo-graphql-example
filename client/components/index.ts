export { Layout } from './Layout/Layout';
export { Wrapper } from './Wrapper/Wrapper';
export { Box } from './Box/Box';
export { BoxText } from './BoxText/BoxText';
export { LikeCount } from './LikeCount/LikeCount';
export { FilterButton } from './FilterButton/FilterButton';
export { LikeButton } from './LikeButton/LikeButton';
export { ProductPrice } from './ProductPrice/ProductPrice';
export { ProductTitle } from './ProductTitle/ProductTitle';
export { ProductDelivery } from './ProductDelivery/ProductDelivery';
export { ProductImage } from './ProductImage/ProductImage';
export { ProductCard } from './ProductCard/ProductCard';
export { ProductCardContainer } from './ProductCardContainer/ProductCardContainer';
export { ButtonGroup } from './ButtonGroup/ButtonGroup';
