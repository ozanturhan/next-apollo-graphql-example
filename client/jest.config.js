module.exports = {
  testPathIgnorePatterns: ['/.next/', '/node_modules/', '/tests/', '/coverage/'],
  testRegex: '(/__test__/.*|\\.(test|spec))\\.(ts|tsx|js|jsx)$',
  moduleFileExtensions: ['ts', 'tsx', 'js', 'json'],
  moduleNameMapper: {
    '@components': '<rootDir>/components',
    '@queries': '<rootDir>/queries',
    '\\.(css)$': '<rootDir>/__mocks__/styleMock.js',
    '\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$':
      '<rootDir>/__mocks__/fileMock.js',
  },
  transform: {
    '^.+\\.(js|jsx|ts|tsx)$': 'babel-jest',
  },
  transformIgnorePatterns: ['<rootDir>/node_modules/'],
  setupFiles: ['./setupJest.ts'],
};
