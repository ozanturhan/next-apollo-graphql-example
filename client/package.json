{
  "name": "client",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "format": "prettier --write './**/*.{js,jsx,ts,tsx,css,md,json}' --config ./.prettierrc",
    "test": "jest",
    "test:watch": "jest --watch",
    "test:coverage": "jest --coverage"
  },
  "dependencies": {
    "@apollo/client": "^3.2.5",
    "enzyme": "^3.11.0",
    "enzyme-adapter-react-16": "^1.15.5",
    "graphql": "^15.3.0",
    "next": "9.5.5",
    "next-images": "^1.6.0",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "styled-components": "^5.2.0"
  },
  "devDependencies": {
    "@babel/core": "^7.12.3",
    "@babel/plugin-proposal-class-properties": "^7.12.1",
    "@babel/plugin-syntax-dynamic-import": "^7.8.3",
    "@babel/plugin-transform-modules-commonjs": "^7.12.1",
    "@babel/plugin-transform-reserved-words": "^7.12.1",
    "@babel/plugin-transform-runtime": "^7.12.1",
    "@babel/plugin-transform-typescript": "^7.12.1",
    "@babel/preset-env": "^7.12.1",
    "@babel/preset-flow": "^7.12.1",
    "@babel/preset-react": "^7.12.1",
    "@babel/preset-typescript": "^7.12.1",
    "@testing-library/react": "^11.1.0",
    "@testing-library/react-hooks": "^3.4.2",
    "@types/jest": "^26.0.15",
    "@types/node": "^14.14.1",
    "@types/react": "^16.9.53",
    "@typescript-eslint/eslint-plugin": "^4.5.0",
    "@typescript-eslint/parser": "^4.5.0",
    "babel-jest": "^26.6.0",
    "babel-loader": "^8.1.0",
    "babel-plugin-inline-react-svg": "^1.1.1",
    "babel-plugin-styled-components": "^1.11.1",
    "babel-plugin-transform-es2015-modules-commonjs": "^6.26.2",
    "babel-preset-react-app": "^9.1.2",
    "jest": "^26.6.0",
    "jest-styled-components": "^7.0.3",
    "prettier": "^2.1.2",
    "react-test-renderer": "^17.0.0",
    "ts-jest": "^26.4.1",
    "typescript": "^4.0.3"
  }
}
