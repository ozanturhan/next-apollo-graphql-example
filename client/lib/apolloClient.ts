import { ApolloClient, InMemoryCache, NormalizedCacheObject } from '@apollo/client';
import { useMemo } from 'react';

let apolloClient;

function createApolloClient(): ApolloClient<NormalizedCacheObject> {
  return new ApolloClient({
    ssrMode: typeof window === 'undefined',
    uri: 'http://localhost:4000/graphql',
    cache: new InMemoryCache(),
  });
}

export function initializeApollo(initialState: {} | null = null) {
  const _apolloClient = apolloClient ?? createApolloClient();

  if (initialState) {
    const existingCache = _apolloClient.extract();
    _apolloClient.cache.restore({ ...existingCache, ...initialState });
  }

  if (typeof window === 'undefined') return _apolloClient;
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function useApollo(initialState) {
  return useMemo(() => initializeApollo(initialState), [initialState]);
}
