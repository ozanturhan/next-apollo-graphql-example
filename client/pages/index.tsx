import React, {useEffect, useReducer, useState} from 'react';
import {useQuery} from '@apollo/client';
import {initializeApollo} from '../lib/apolloClient';
import {PRODUCTS_QUERY} from '@queries';
import {ButtonGroup, FilterButton, Layout, LikeCount, ProductCard, ProductCardContainer, Wrapper} from '@components';
import {Product} from '../lib/product';

const initialState: Product[] | null = null;

const reducer = (state, action) => {
  switch (action.type) {
    case 'toggle':
      return state.map((item) => (item.id === action.id ? { ...item, isLiked: !item.isLiked } : item));
    case 'fill':
      return action.products;
    default:
      return state;
  }
};

const FILTER_MAP = {
  All: () => true,
  Liked: (item) => item.isLiked,
};

function Home() {
  const { data } = useQuery(PRODUCTS_QUERY);
  const [filterType, setFilterType] = useState('All');
  const [state, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    dispatch({ type: 'fill', products: data?.products });
  }, [data]);

  useEffect(() => {
    const likeCount = state?.filter((item) => item.isLiked)?.length;
    if (likeCount === 0 && filterType === 'Liked') {
      handleFilter(false)
    }
  }, [state]);

  const products: Product[] = state || data?.products;

  const handleFilter = (value) => {
    setFilterType(value ? 'Liked' : 'All');
  };

  const likeCount = state?.filter((item) => item.isLiked)?.length;

  return (
    <Layout>
      <Wrapper>
        <ButtonGroup>
          <LikeCount count={likeCount} />

          <FilterButton disabled={likeCount === 0} onToggle={handleFilter} isToggle={filterType === 'Liked'}>
            Beğendiklerim
          </FilterButton>
        </ButtonGroup>

        <ProductCardContainer title="İlginizi Çekebilecek Ürünler">
          {products?.filter(FILTER_MAP[filterType]).map((item) => (
            <ProductCard
              data-testid="product"
              key={item.id}
              item={item}
              onToggleLike={() => dispatch({ type: 'toggle', id: item.id })}
            />
          ))}
        </ProductCardContainer>
      </Wrapper>
    </Layout>
  );
}

export async function getServerSideProps() {
  const apolloClient = initializeApollo();

  await apolloClient.query({
    query: PRODUCTS_QUERY,
  });

  return {
    props: {
      initialApolloState: apolloClient.cache.extract(),
      revalidate: 1,
    },
  };
}

export default Home;
