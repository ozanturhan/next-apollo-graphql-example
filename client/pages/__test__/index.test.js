import React from 'react';
import Home, { getServerSideProps } from '../index';
import { MockedProvider } from '@apollo/client/testing';
import { PRODUCTS_QUERY } from '@queries';
import { initializeApollo } from '../../lib/apolloClient';
import { act, render } from '@testing-library/react';
import { mount } from 'enzyme';

jest.mock('../../lib/apolloClient.ts');

const mockResult = {
  data: {
    products: [
      {
        id: 1,
        image: 'http://localhost:4001/image/256/iphone_11.jpg',
        isLiked: false,
        name: 'Apple iPhone 11 128 GB Pro Cep Telefonu Apple Türkiye Garantili',
        price: 1250,
        cargoType: 'FREE',
      },
      {
        id: 2,
        image: 'http://localhost:4001/image/256/iphone_11.jpg',
        isLiked: true,
        name: 'Apple iPhone 12 128 GB Pro Cep Telefonu Apple Türkiye Garantili',
        price: 1250,
        cargoType: 'FREE',
      },
    ],
  },
};

const mocks = [
  {
    request: {
      query: PRODUCTS_QUERY,
    },
    result: mockResult,
  },
];

describe('Index', () => {
  it('should be rendered correctly', async () => {
    // Arrange && Act
    const tree = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Home />
      </MockedProvider>,
    ).container;

    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

    // Assertion
    expect(tree).toMatchSnapshot();
  });

  it('should be filtered when clicking filter button', async () => {
    // Arrange
    const wrapper = mount(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Home />
      </MockedProvider>,
    );

    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });

    // Act
    wrapper.find('button').at(0).simulate('click');

    // Assertion
    expect(wrapper.find("[data-testid='product']")).toHaveLength(1);

    // Act
    wrapper.find('button').at(0).simulate('click');

    // Assertion
    expect(wrapper.find("[data-testid='product']")).toHaveLength(2);
  });

  it('should get products form server', async () => {
    // Arrange
    const query = jest.fn();

    initializeApollo.mockImplementation(() => ({
      query,
      cache: {
        extract: () => mockResult,
        restore: jest.fn(),
      },
    }));

    // Act
    const response = await getServerSideProps();

    // Assertion
    expect(response).toEqual(
      expect.objectContaining({
        props: {
          initialApolloState: mockResult,
          revalidate: 1,
        },
      }),
    );
  });
});
