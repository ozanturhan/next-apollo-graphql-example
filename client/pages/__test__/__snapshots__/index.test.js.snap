// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index should be rendered correctly 1`] = `
<div>
  <div
    class="Layout-b1bw1s-0 eoVKzC"
  >
    <div
      class="Wrapper-e1775g-0 cCgOfN"
    >
      <div
        class="ButtonGroup-sc-5v2puq-0 iEXrzR"
      >
        <div
          class="Box-sc-1k72gju-0 lfkyrw"
        >
          <svg
            class="LikeCount__Heart-sc-1c2au0a-0 gUSNkV"
            viewBox="0 0 391.837 391.837"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M285.257 35.528c58.743.286 106.294 47.836 106.58 106.58 0 107.624-195.918 214.204-195.918 214.204S0 248.165 0 142.108c0-58.862 47.717-106.58 106.58-106.58a105.534 105.534 0 0 1 89.339 48.065 106.578 106.578 0 0 1 89.338-48.065z"
              fill="#d7443e"
            />
          </svg>
          <span
            class="BoxText-o9bdxb-0 hPeSsh"
          >
            1
             ürün
          </span>
        </div>
        <div
          class="Box-sc-1k72gju-0 fdNxIG"
        >
          <button
            class="FilterButton__Button-sc-1bx921t-0 ixlYOT"
          >
            Beğendiklerim
          </button>
        </div>
      </div>
      <div
        class="ProductCardContainer__Container-sc-16wksml-0 bTJrUm"
      >
        <h2>
          İlginizi Çekebilecek Ürünler
        </h2>
        <div
          class="ProductCardContainer__GridContainer-sc-16wksml-1 BjqaQ"
        >
          <div
            class="ProductCard__Container-sc-1hvs6bu-0 jvmzUk"
          >
            <button
              class="LikeButton__Button-eur3kr-0 fQYiDT"
            >
              <svg
                class="LikeButton__Heart-eur3kr-1 cJjbGJ"
                viewBox="0 0 391.837 391.837"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M285.257 35.528c58.743.286 106.294 47.836 106.58 106.58 0 107.624-195.918 214.204-195.918 214.204S0 248.165 0 142.108c0-58.862 47.717-106.58 106.58-106.58a105.534 105.534 0 0 1 89.339 48.065 106.578 106.578 0 0 1 89.338-48.065z"
                  fill="#d7443e"
                />
              </svg>
            </button>
            <a
              href="http://localhost:4001/image/256/iphone_11.jpg"
            >
              <span
                class="ProductImage__Container-sc-13be745-0 kWuzFc"
              >
                <img
                  alt="http://localhost:4001/image/256/iphone_11.jpg"
                  src="http://localhost:4001/image/256/iphone_11.jpg"
                />
              </span>
              <span
                class="ProductTitle-sc-12u2ti2-0 dnSkZP"
                data-testid="product-title"
              >
                Apple iPhone 11 128 GB Pro Cep Telefonu Apple Türkiye Garantili
              </span>
              <span
                class="ProductPrice-o0iam0-0 fErfTM"
              >
                1250
              </span>
              <span
                class="ProductDelivery__ProductDeliveryContainer-sc-11e1vbf-0 cTnboe"
              >
                <svg
                  class="ProductDelivery__Icon-sc-11e1vbf-1 iPMphp"
                  viewBox="0 -104 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M365 3.016c44.39 10.3 61.371 36.257 82.66 71.984H365zM105.578 78c8.692 0 15 4.195 15 14 0 8.27-6.691 14.977-14.957 15H15c-8.285 0-15 6.719-15 15 0 8.285 6.715 15 15 15h135c8.363 0 15.059 6.71 15.059 15 0 8.285-6.715 15-15 15H15c-8.285 0-15 6.715-15 15s6.715 15 15 15h33v45c0 8.285 6.715 15 15 15h30.152c5.375 26.477 28.77 46 56.348 46s50.973-19.523 56.348-46h152.304c5.375 26.477 28.77 46 56.348 46s50.973-19.523 56.348-46H497c8.285 0 15-6.715 15-15v-90c0-44.012-46.422-46.934-46.465-47H350c-8.285 0-15-6.715-15-15V0H63c-8.285 0-15 6.715-15 15v33H30c-8.285 0-15 6.715-15 15s6.715 15 15 15zm328.367 148.055c10.739 10.738 10.739 28.156 0 38.894C416.672 282.223 387 269.934 387 245.5c0-24.43 29.672-36.719 46.945-19.445zm-265 0c10.739 10.738 10.739 28.156 0 38.894C151.672 282.223 122 269.934 122 245.5c0-24.43 29.672-36.719 46.945-19.445zm0 0"
                    fill="#1CB55E"
                    fill-rule="evenodd"
                  />
                </svg>
                 
                Ücretsiz
              </span>
            </a>
          </div>
          <div
            class="ProductCard__Container-sc-1hvs6bu-0 jvmzUk"
          >
            <button
              class="LikeButton__Button-eur3kr-0 jIZdPW"
            >
              <svg
                class="LikeButton__Heart-eur3kr-1 cJjbGJ"
                viewBox="0 0 391.837 391.837"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M285.257 35.528c58.743.286 106.294 47.836 106.58 106.58 0 107.624-195.918 214.204-195.918 214.204S0 248.165 0 142.108c0-58.862 47.717-106.58 106.58-106.58a105.534 105.534 0 0 1 89.339 48.065 106.578 106.578 0 0 1 89.338-48.065z"
                  fill="#d7443e"
                />
              </svg>
            </button>
            <a
              href="http://localhost:4001/image/256/iphone_11.jpg"
            >
              <span
                class="ProductImage__Container-sc-13be745-0 kWuzFc"
              >
                <img
                  alt="http://localhost:4001/image/256/iphone_11.jpg"
                  src="http://localhost:4001/image/256/iphone_11.jpg"
                />
              </span>
              <span
                class="ProductTitle-sc-12u2ti2-0 dnSkZP"
                data-testid="product-title"
              >
                Apple iPhone 12 128 GB Pro Cep Telefonu Apple Türkiye Garantili
              </span>
              <span
                class="ProductPrice-o0iam0-0 fErfTM"
              >
                1250
              </span>
              <span
                class="ProductDelivery__ProductDeliveryContainer-sc-11e1vbf-0 cTnboe"
              >
                <svg
                  class="ProductDelivery__Icon-sc-11e1vbf-1 iPMphp"
                  viewBox="0 -104 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M365 3.016c44.39 10.3 61.371 36.257 82.66 71.984H365zM105.578 78c8.692 0 15 4.195 15 14 0 8.27-6.691 14.977-14.957 15H15c-8.285 0-15 6.719-15 15 0 8.285 6.715 15 15 15h135c8.363 0 15.059 6.71 15.059 15 0 8.285-6.715 15-15 15H15c-8.285 0-15 6.715-15 15s6.715 15 15 15h33v45c0 8.285 6.715 15 15 15h30.152c5.375 26.477 28.77 46 56.348 46s50.973-19.523 56.348-46h152.304c5.375 26.477 28.77 46 56.348 46s50.973-19.523 56.348-46H497c8.285 0 15-6.715 15-15v-90c0-44.012-46.422-46.934-46.465-47H350c-8.285 0-15-6.715-15-15V0H63c-8.285 0-15 6.715-15 15v33H30c-8.285 0-15 6.715-15 15s6.715 15 15 15zm328.367 148.055c10.739 10.738 10.739 28.156 0 38.894C416.672 282.223 387 269.934 387 245.5c0-24.43 29.672-36.719 46.945-19.445zm-265 0c10.739 10.738 10.739 28.156 0 38.894C151.672 282.223 122 269.934 122 245.5c0-24.43 29.672-36.719 46.945-19.445zm0 0"
                    fill="#1CB55E"
                    fill-rule="evenodd"
                  />
                </svg>
                 
                Ücretsiz
              </span>
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
